### PR TITLE
Alternative implementation for JSMA

### PR DESCRIPTION
**Pull request by zHiOoO**
I know there is currently another open pull request for the JSMA, but I just want to share the work I have done and hopefully it can help too.

My motivation for the change is to improve the efficiency to generate adversarial examples for JSMA. With the change, the time needed for generation should be greatly reduced. I haven't done a precise measurement, but it roughly takes about 15 sec to generate 500 adversarial samples with GPU.

The change tries to keeps the same interface of the original implementation, so that no modification is needed for tutorial.

I have run the tests in the test_tf and the tutorial. I think there is something wrong with the "test_generate_np_targeted_gives_adversarial_example" test. I tried to fix it, but I can't find a good way to do so. Maybe someone can comment on what the test is trying to do.

Any comments and suggestions are appreciated.

### CODE REVIEW

**Comment by npapernot** [comment]
Thanks for the pull request, in the future it would be best to discuss this on the cleverhans-dev@ mailing list to avoid duplicate PRs. I do appreciate you sharing your work.

Let us see if it makes sense to keep #275 given that this PR has been open. @gongzhitaao are you planning to complete your PR soon, or should we close it and all work towards this current PR being merged (#319)?

File: cleverhans/attacks_tf.py [hunk 1]
```diff
@@ -257,151 +184,138 @@ def jacobian_graph(predictions, x, nb_classes):
...
+
+        # Create a mask to only keep features that match conditions
+        if increase:
+            scores_mask = ((target_sum > 0) & (other_sum < 0))
+        else:
+            scores_mask = ((target_sum < 0) & (other_sum > 0))
 
-        # Update loop variables
-        iteration = iteration + 1
+        # Create a 2D numpy array of scores for each pair of candidate features
```
Comment (gongzhitaao) [inline cleverhans/attacks_tf.py]: Quick question: suppose we have an N by N image and we need to find the best pair to perturb, according to this code, we need a matrix of N^2 by N^2, am I understanding it right? In this case, suppose we want to perturb an image of size100x100, then we need approx. 2G memory.

**Comment by gongzhitaao** [comment]
I'm OK either way.  I quickly went through the implementation. As far as I could see, for smaller image size, this solution should be faster than mine since the graph is simpler and more vectorized.
@zHiOoO Have you tested 256x256 images from ImageNet?

File: cleverhans/attacks_tf.py (hunk 1 above)
Comment (zHiOoO) [reply cleverhans/attacks_tf.py]: Yes, you are right. And it will take more memory if the image has multiple channels.

**Comment by zHiOoO** [comment]
@gongzhitaao No, I haven't tried using images with 256x256. You have a very good point. What I have done is a direct port from the original implementation. It works well for simple dataset like MNIST, but it does has memory issue for datasets with larger image size.

**Comment by gongzhitaao** [comment]
@npapernot I think we could close my PR #275 and work toward this submission.
@zHiOoO Your code has some good idea that I want to integrate into my own implementation.  Thank you for the PR :smiley:

**Comment by npapernot** [comment]
Thanks for taking the time to look at this PR @gongzhitaao , I will close #275 for now. Hope that you are able to build on / contribute to this PR with your implementation ideas.

File: cleverhans/attacks_tf.py [hunk 2]
```diff
@@ -257,151 +184,138 @@ def jacobian_graph(predictions, x, nb_classes):
...
-        current = utils_tf.model_argmax(sess, x, predictions,
-                                        adv_x_original_shape, feed=feed)
+        target_class = tf.reshape(tf.transpose(y_in, perm=[1, 0]),
+                                  shape=[nb_classes, -1, 1])
+        other_classes = tf.cast(tf.not_equal(target_class, 1), tf.float32)
+
+        grads_target = tf.reduce_sum(grads * target_class, axis=0)
+        grads_other = tf.reduce_sum(grads * other_classes, axis=0)
+
+        # Remove the already-used input features from the search space
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: Why not just set the values of `grads_target` and `grads_other` to 0 for the features in domain_in? These way they will not match the `scores_mask` conditions below.

File: cleverhans/attacks_tf.py [hunk 3]
```diff
@@ -257,151 +184,138 @@ def jacobian_graph(predictions, x, nb_classes):
...
+        p1_one_hot = tf.one_hot(p1, depth=nb_features)
+        p2_one_hot = tf.one_hot(p2, depth=nb_features)
 
-    # Report success when the adversarial example is misclassified in the
-    # target class
-    if current == target:
-        return np.reshape(adv_x, original_shape), 1, percent_perturbed
-    else:
-        return np.reshape(adv_x, original_shape), 0, percent_perturbed
+        # Check if more modification is needed for each samples
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: "for each sample"

File: cleverhans/attacks_tf.py [hunk 4]
```diff
@@ -257,151 +184,138 @@ def jacobian_graph(predictions, x, nb_classes):
...
+        mod_done = tf.equal(tf.reduce_sum(y_in * preds_onehot, axis=1), 0)
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: Should this be called `mod_not_done` instead?

File: cleverhans/attacks.py [hunk 5]
```diff
@@ -486,39 +486,35 @@ def generate(self, x, **kwargs):
...
-            wrap = tf.py_func(jsma_wrap, [x], tf.float32)
+                for i in xrange(nb_s):
+                    result[i, :] = np.roll(result[i, :],
+                                           randint(1, nb_classes-1))
 
-        return wrap
+                return result
+
+            labels, nb_classes = self.get_or_guess_labels(x, kwargs)
+            self.y_target = tf.py_func(random_targets, [labels], tf.float32)
```
[the remark this answers is not in the document]
Comment (zHiOoO) [reply cleverhans/attacks.py]: I would love to avoid using py_func here, but I couldn't think of a good way to do this with pure tf functions.

File: cleverhans/attacks.py [hunk 6]
```diff
@@ -486,39 +486,35 @@ def generate(self, x, **kwargs):
...
+            self.y_target.set_shape([None, nb_classes])
```
[the remark this answers is not in the document]
Comment (zHiOoO) [reply cleverhans/attacks.py]: This is necessary because the output of py_func has unknown shape.

File: cleverhans/attacks_tf.py [hunk 7]
```diff
@@ -133,79 +133,6 @@ def vatm(model, x, logits, eps, num_iterations=1, xi=1e-6,
         return adv_x
 
 
-def apply_perturbations(i, j, X, increase, theta, clip_min, clip_max):
```
[the remark this answers is not in the document]
Comment (zHiOoO) [reply cleverhans/attacks_tf.py]: Sure. I can add them back.

File: cleverhans/attacks_tf.py [hunk 8]
```diff
@@ -257,151 +184,138 @@ def jacobian_graph(predictions, x, nb_classes):
     return list_derivatives
 
 
-def jsma(sess, x, predictions, grads, sample, target, theta, gamma, clip_min,
-         clip_max, feed=None):
+def jsma(x, y_target, model, theta, gamma, clip_min, clip_max):
     """
     TensorFlow implementation of the JSMA (see https://arxiv.org/abs/1511.07528
     for details about the algorithm design choices).
-    :param sess: TF session
```
[the remark this answers is not in the document]
Comment (zHiOoO) [reply cleverhans/attacks_tf.py]: Because the function no longer requires this input.

File: cleverhans/attacks_tf.py [hunk 9]
```diff
@@ -257,151 +184,138 @@ def jacobian_graph(predictions, x, nb_classes):
...
+        # reaches the maximum iterations
+        return tf.logical_and(tf.less(i_in, max_iters), cond_in)
 
-    _logger.debug("Starting JSMA attack up to {} iterations".format(max_iters))
-    # Repeat this main loop until we have achieved misclassification
-    while (current != target and iteration < max_iters and
-           len(search_domain) > 1):
-        # Reshape the adversarial example
-        adv_x_original_shape = np.reshape(adv_x, original_shape)
+    def body(x_in, y_in, domain_in, i_in, cond_in):
```
[the remark this answers is not in the document]
Comment (zHiOoO) [reply cleverhans/attacks_tf.py]: Will do.

File: cleverhans/attacks_tf.py [hunk 10]
```diff
@@ -257,151 +184,138 @@ def jacobian_graph(predictions, x, nb_classes):
...
-            grads_target, grads_others, search_domain, increase)
-
-        # Apply the perturbation to the two input features selected previously
-        adv_x = apply_perturbations(
-            i, j, adv_x, increase, theta, clip_min, clip_max)
-
-        # Update our current prediction by querying the model
-        current = utils_tf.model_argmax(sess, x, predictions,
-                                        adv_x_original_shape, feed=feed)
+        target_class = tf.reshape(tf.transpose(y_in, perm=[1, 0]),
```
[the remark this answers is not in the document]
Comment (zHiOoO) [reply cleverhans/attacks_tf.py]: Will do

File: cleverhans/attacks_tf.py (hunk 2 above)
Comment (zHiOoO) [reply cleverhans/attacks_tf.py]: The algorithm is picking pairs of pixels/features, so setting one pixel/feature to 0 doesn't prevent it from being picked again later. That's why their values need to be set based on the maximum value from the gradients.

File: cleverhans/attacks_tf.py (hunk 3 above)
Comment (zHiOoO) [reply cleverhans/attacks_tf.py]: Will fix

File: cleverhans/attacks_tf.py (hunk 4 above)
Comment (zHiOoO) [reply cleverhans/attacks_tf.py]: Good catch. Will fix.

**Comment by goodfeli** [comment]
@npapernot I think it's ready for you to re-review

**Comment by npapernot** [comment]
Thanks, this is almost ready to merge! just two things would be nice to take care of:

* solve the conflict that the branch has w/ the master branch
* change the JSMA tutorial in `cleverhans_tutorials` to use the symbolic implementation by default (this will also make sure that the behavior is consistent with the previous implementation)

Thanks again

**Comment by zHiOoO** [comment]
@npapernot The conflict is resolved. I have also added a new parameter for JSMA to allow switching between two implementations.

**Comment by npapernot** [comment]
Thanks, I'll try a few things on my machine and then merge
